lib/core/src/sync/mod.rs

```diff
@@ -208,6 +227,124 @@ impl SyncService {
 
         Ok(())
     }
+
+    async fn push(&self) -> Result<()> {
+        let outgoing_details = self.persister.get_sync_outgoing_details()?;
+
+        for SyncOutgoingDetails {
+            record_id,
+            record_type,
+            ..
+        } in outgoing_details
+        {
+            // Step 1: Get the sync state, ensure it exists
+            let Some(sync_state) = unwrap_or_continue!(
+                self.persister.get_sync_state_by_record_id(&record_id),
+                "Could not retrieve sync state: {}"
+            ) else {
+                log::debug!("Cannot sync record without any sync state. Skipping.");
```

Comment:
When clients upgrade the sdk how would we push their existing swaps (first revision)?
One way is to add a migration step to add sync state for all existing swaps. The other option is to include get_sync_outgoing_details all swaps that without sync details.
I somehow like the second option because:
1. It doesn't need any migration step
2. We don't need to add sync_state on every time we adda swap and it helps us also separate the sync logic from the business logic.

Reply:
> One way is to add a migration step to add sync state for all existing swaps

After some thought/experimentation, I figure it's the _only way_ to do so with the current DB structure. The push process _requires_ a swap state in order to work. Let me explain more clearly why:
- User creates a swap, persister method `insert_X_swap` is called
- `insert_X_swap` calls `commit_new`. This creates only the SyncOutgoingDetails, with record_id, record_type, commit_type and updated_fields (None).
- On the next sync event loop, fetch all `SyncOutgoingDetails`. If we go by those alone, then we will have a list of record_ids, but no way to associate them to the given swap (to then load the swap data, and create the outgoing record).

I understand your point about creating those values on the fly (record_id and record_revision), but that cannot be done for the swap_id (data_id).

Two ways to go about this:
1. Add `data_id` to `sync_outgoing` as well
2. Ensure both swap_state and sync_outgoing exist at the time of push, else skip

I'm against neither of those, and I agree it would spare a (large) db insert for new clients with many local swaps if we went for the 1st option.